File: shared/src/types/booking.ts
// ─── Vehicle Types ─────────────────────────────────────────────────────────

export interface FareItem {
  label: string;
  amount: number;
}

export interface FareBreakdownData {
  // Identifiers
  pricing_rule_version_id?: string;
  pricing_version_tag?: string;
  base_rate_id?: string;
  city_id?: string | null;
  city_name?: string;

  // Trip metadata
  ride_type: string;
  vehicle_category: string;
  is_round_trip?: boolean;
  effective_distance_km?: number;
  currency?: string;
  calculated_at?: string;

  // Legacy compat (old field names still accepted)
  distance_km?: number;
  duration_min?: number;

  // Core fare components
  base_fare: number;
  distance_fare: number;
  time_fare?: number;        // new name (was duration_fare)
  duration_fare?: number;    // old name — kept for backwards compat
  subtotal_raw?: number;
  minimum_fare?: number;
  minimum_fare_applied?: boolean;

  // Multipliers
  peak_multiplier?: number;
  peak_amount?: number;
  night_multiplier?: number;
  night_amount?: number;
  holiday_multiplier?: number;
  holiday_amount?: number;
  surge_multiplier?: number;
  surge_amount?: number;     // new name
  surge_charge?: number;     // old name — kept for backwards compat
  weather_multiplier?: number;
  weather_amount?: number;

  // After multipliers
  subtotal_after_multipliers?: number;

  // Incidentals
  waiting_charge: number;
  toll_charges: number;
  parking_charges?: number;
  airport_pickup_charge?: number;
  airport_drop_charge?: number;
  driver_allowance: number;
  state_border_charge?: number;
  night_charge?: number;     // old name — kept for backwards compat

  // After incidentals
  subtotal_after_incidentals?: number;

  // Discounts
  coupon_code?: string | null;
  coupon_discount: number;
  wallet_deduction?: number;
  total_discounts?: number;
  subtotal_after_discounts?: number;
  subtotal_before_discounts?: number;

  // Taxes & fees
  gst_percentage?: number;
  gst_amount?: number;
  insurance_fee?: number;
  platform_fee?: number;
  state_tax?: number;        // old name — kept for backwards compat

  // Total
  total_fare: number;
  final_fare_unrounded?: number;

  // Breakdown summary
  breakdown?: Record<string, number>;

  // Legacy compat
  is_peak?: boolean;
}

export interface Vehicle {
  category: string;
  display_name: string;
  icon: string;
  seats: number;
  luggage: string;
  comfort: string;
  fuel_type: string;
  cancellation_policy: string;
  eta_min: number;
  eta_max: number;
  eta_display: string;
  fare: number;
  fare_breakdown: FareBreakdownData;
  ride_types: string[];
}

// ─── Booking Preferences ───────────────────────────────────────────────────

export interface BookingPreferences {
  preferred_language: string;
  air_conditioning: boolean;
  pet_friendly: boolean;
  wheelchair_accessible: boolean;
  female_driver_preferred: boolean;
  child_seat: boolean;
  silent_ride: boolean;
  music_preference: string | null;
  special_instructions: string | null;
}

export const defaultPreferences: BookingPreferences = {
  preferred_language: 'en',
  air_conditioning: true,
  pet_friendly: false,
  wheelchair_accessible: false,
  female_driver_preferred: false,
  child_seat: false,
  silent_ride: false,
  music_preference: null,
  special_instructions: null,
};

// ─── Coupon ────────────────────────────────────────────────────────────────

export interface CouponResult {
  valid: boolean;
  code?: string;
  description?: string;
  discount_type?: string;
  discount_amount?: number;
  final_fare?: number;
  error?: string;
}

// ─── Booking ──────────────────────────────────────────────────────────────

export type TripType = 'NOW' | 'SCHEDULED' | 'ROUND_TRIP';
export type BookingStatus = 'PENDING' | 'SEARCHING' | 'DRIVER_ASSIGNED' | 'IN_PROGRESS' | 'COMPLETED' | 'CANCELLED';

export interface Booking {
  booking_ref: string;
  status: BookingStatus;
  ride_type: string;
  vehicle_category: string;
  pickup_address: string;
  destination_address: string;
  distance_km: number;
  duration_min: number;
  scheduled_at?: string;
  payment_method: string;
  created_at: string;
}

File: backend/app/schemas/booking.py
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, UUID4, Field
from datetime import datetime
from app.models.booking import RideType, TripType, BookingStatus, VehicleCategory, CouponType

class FareEstimateBase(BaseModel):
    ride_type: RideType
    vehicle_category: VehicleCategory
    distance_km: float
    duration_min: float

class FareEstimateRequest(BaseModel):
    ride_type: RideType
    pickup_lat: float
    pickup_lon: float
    destination_lat: Optional[float] = None
    destination_lon: Optional[float] = None
    distance_km: Optional[float] = None
    duration_min: Optional[float] = None
    is_round_trip: Optional[bool] = False


class FareEstimateResponse(FareEstimateBase):
    id: UUID4
    base_fare: float
    distance_fare: float
    duration_fare: float
    surge_multiplier: float
    surge_charge: float
    toll_charges: float
    night_charge: float
    driver_allowance: float
    state_tax: float
    coupon_discount: float
    total_fare: float
    breakdown: Optional[Dict[str, Any]]
    created_at: datetime
    
    class Config:
        from_attributes = True

class CouponValidateRequest(BaseModel):
    code: str
    ride_type: Optional[RideType] = None
    fare_amount: float

class CouponValidateResponse(BaseModel):
    valid: bool
    discount_amount: float
    message: Optional[str] = None

class BookingPreferenceSchema(BaseModel):
    preferred_language: Optional[str] = "en"
    air_conditioning: Optional[bool] = True
    pet_friendly: Optional[bool] = False
    wheelchair_accessible: Optional[bool] = False
    female_driver_preferred: Optional[bool] = False
    child_seat: Optional[bool] = False
    silent_ride: Optional[bool] = False
    music_preference: Optional[str] = None
    special_instructions: Optional[str] = None

class BookingCreateRequest(BaseModel):
    ride_type: RideType
    trip_type: Optional[TripType] = TripType.NOW
    vehicle_category: VehicleCategory
    pickup_lat: float
    pickup_lon: float
    pickup_address: str
    destination_lat: float
    destination_lon: float
    destination_address: str
    distance_km: Optional[float] = None
    duration_min: Optional[float] = None
    route_geometry: Optional[Dict[str, Any]] = None
    fare_estimate_id: Optional[UUID4] = None
    coupon_code: Optional[str] = None
    payment_method: Optional[str] = "CASH"
    scheduled_at: Optional[datetime] = None
    return_at: Optional[datetime] = None
    preferences: Optional[BookingPreferenceSchema] = None
    idempotency_key: Optional[str] = None
    fare_breakdown: Optional[Dict[str, Any]] = None

class BookingResponse(BaseModel):
    id: UUID4
    booking_ref: str
    rider_id: UUID4
    ride_type: RideType
    trip_type: TripType
    vehicle_category: VehicleCategory
    status: BookingStatus
    pickup_address: str
    destination_address: str
    payment_method: str
    created_at: datetime
    
    class Config:
        from_attributes = True
